lib/tasks/sitemap.rake

```diff
@@ -8,8 +10,35 @@ namespace :sitemap do
 
     output_directory = File.join(PROJECT_ROOT, "public")
     sitemap = Sitemap.new(output_directory)
-    sitemap.generate(SearchConfig.default_instance)
+    sitemap.generate_and_replace(SearchConfig.default_instance)
 
     sitemap.cleanup
   end
+
+  desc "Generate new sitemap files and upload to S3"
+  task :generate_and_upload do
+    @bucket_name = ENV['AWS_S3_BUCKET_NAME']
+    raise "Missing required AWS_S3_BUCKET_NAME" if @bucket_name.nil?
+
+    Dir.mktmpdir do |output_directory|
+      filenames = Sitemap.new(output_directory).generate(SearchConfig.default_instance)
+
+      filenames[:sitemaps].each do |filename, link_filename|
+        upload_to_s3(
+          "#{output_directory}/#{Sitemap::SUB_DIRECTORY}/#{filename}",
+          "#{Sitemap::SUB_DIRECTORY}/#{link_filename}"
+        )
+      end
+
+      upload_to_s3(
+        "#{output_directory}/#{Sitemap::SUB_DIRECTORY}/#{filenames[:index]}",
+        "sitemap.xml"
+      )
+    end
+  end
+
+  def upload_to_s3(source, target)
+    o = Aws::S3::Object.new(bucket_name: @bucket_name, key: target)
+    raise "Failed to upload sitemap file '#{source}' as '#{target}'" unless o.upload_file(source)
```

Comment:
Can `upload_file` raise an error to explain why the upload failed?

Reply:
I think so.  Earlier I had a problem where I got the source file path wrong, and got an error to that effect.  Unsure about AWS-side errors, but I'd assume so.

Reply:
It would be great to capture that, especially as I guess we'll want to have an Icinga alert should this upload fail. I can see myself retrying it if that happens, since it'll be hard to debug.

Reply:
It throws at least one type of exception - possibly more.  But it also returns a boolean.  Maybe every error will raise an exception, so it'll only ever return `true`, but maybe not.